### PR TITLE
allow coder login for WSL

### PR DESCRIPTION
**Pull request by arminaaki**
## Issue
`coder login` fails to open user's browser while running from [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/).
```
$ coder login <coder_url>
/usr/bin/xdg-open: 869: www-browser: not found
/usr/bin/xdg-open: 869: links2: not found
/usr/bin/xdg-open: 869: elinks: not found
/usr/bin/xdg-open: 869: links: not found
/usr/bin/xdg-open: 869: lynx: not found
/usr/bin/xdg-open: 869: w3m: not found
```

## Changes
* allow `coder login` to work while running from WSL.
* Detect if running in WSL by identifying `microsoft` in  `/proc/version` file.
```
$ cat /proc/version
Linux version ....-Microsoft (Microsoft@Microsoft.com) (gcc version ... (GCC) ) #...-Microsoft ....
```

### CODE REVIEW

**Review by vapurrmaid — COMMENTED** [review]
Code looks great to me, appreciate the extra log! Thanks for the contribution! 👍🏻

File: internal/cmd/login.go [hunk 1]
```diff
@@ -113,3 +117,36 @@ func pingAPI(ctx context.Context, workspaceURL *url.URL, token string) error {
 	}
 	return nil
 }
+
+// isWSL determines if coder-cli is running within Windows Subsystem for Linux
+func isWSL() (bool, error) {
+	if runtime.GOOS == "darwin" || runtime.GOOS == "windows" {
```
Comment (vapurrmaid) [inline internal/cmd/login.go:123]: it looks like we should make use of constants that are already defined:

https://github.com/cdr/coder-cli/blob/582f213d183be7135e2fcf0e6f839ea320e6c2f3/internal/cmd/update.go#L37

we should add `goosDarwin` as a const as well

File: internal/cmd/login.go [hunk 2]
```diff
@@ -113,3 +117,36 @@ func pingAPI(ctx context.Context, workspaceURL *url.URL, token string) error {
...
+
+	wsl, err := isWSL()
+	if err != nil {
+		return xerrors.Errorf("test running Windows Subsystem for Linux: %w", err)
+	}
+
+	if wsl {
+		cmd = "cmd.exe"
+		args = []string{"/c", "start"}
+		url = strings.Replace(url, "&", "^&", -1)
```
Comment (vapurrmaid) [inline internal/cmd/login.go:146]:
```suggestion
		url = strings.ReplaceAll(url, "&", "^&")
```

File: internal/cmd/login.go [hunk 3]
```diff
@@ -113,3 +117,36 @@ func pingAPI(ctx context.Context, workspaceURL *url.URL, token string) error {
...
+	}
+	return strings.Contains(strings.ToLower(string(data)), "microsoft"), nil
+}
+
+// openURL opens the provided URL via user's default browser
+func openURL(url string) error {
+	var cmd string
+	var args []string
+
+	wsl, err := isWSL()
```
Comment (jawnsy) [inline internal/cmd/login.go:138]: I think this is good, but I do wonder whether we could propose this upstream to pkg/browser

Comment (arminaaki) [reply internal/cmd/login.go:138]: Thank you for the review. I agree, this work should ideally be  part of upstream `pkg/browser`. But Thoughts on moving forward with the work for Coder until I have the patch ready for upstream?

File: internal/cmd/login.go (hunk 1 above)
Comment (arminaaki) [reply internal/cmd/login.go]: Thank you for your review. Updated!

File: internal/cmd/login.go (hunk 2 above)
Comment (arminaaki) [reply internal/cmd/login.go]: Done!

File: internal/cmd/login.go (hunk 1 above)
Comment (arminaaki) [reply internal/cmd/login.go]: This may resolve  → https://github.com/cdr/coder-cli/pull/448

File: internal/cmd/login.go (hunk 3 above)
Comment (jawnsy) [reply internal/cmd/login.go:138]: Yeah, I think it makes sense to make this change now for sure! No objection to it :)

Thanks for your contribution!

**Review by vapurrmaid — APPROVED** [review]

**Pull request merged**